x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanStreamOutput.java

```diff
@@ -130,4 +158,57 @@ public void setTransportVersion(TransportVersion version) {
         delegate.setTransportVersion(version);
         super.setTransportVersion(version);
     }
+
+    /**
+     * Write a {@link Block} as part of the plan.
+     * <p>
+     *     These {@link Block}s are not tacked by {@link BlockFactory} and closing them
+     *     does nothing so they should be small. We do make sure not to send duplicates,
+     *     reusing blocks sent as part of the {@link EsqlConfiguration#tables()} if
+     *     possible, otherwise sending a {@linkplain Block} inline.
+     * </p>
+     */
+    public void writeBlock(Block block) throws IOException {
+        assert block instanceof LongBigArrayBlock == false : "BigArrays not supported because we don't close";
+        assert block instanceof IntBigArrayBlock == false : "BigArrays not supported because we don't close";
+        assert block instanceof DoubleBigArrayBlock == false : "BigArrays not supported because we don't close";
+        assert block instanceof BooleanBigArrayBlock == false : "BigArrays not supported because we don't close";
+        BytesReference key = cachedBlocks.get(block);
+        if (key != null) {
+            key.writeTo(this);
+            return;
+        }
+        writeByte(NEW_BLOCK_KEY);
+        writeVInt(nextCachedBlock);
+        cachedBlocks.put(block, fromPreviousKey(nextCachedBlock));
+        writeNamedWriteable(block);
+        nextCachedBlock++;
+    }
+
+    static final byte NEW_BLOCK_KEY = 0;
+
+    static final byte FROM_PREVIOUS_KEY = 1;
+
+    static BytesReference fromPreviousKey(int id) {
+        try (BytesStreamOutput key = new BytesStreamOutput()) {
+            key.writeByte(FROM_PREVIOUS_KEY);
+            key.writeVInt(id);
+            return key.bytes();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    static final byte FROM_CONFIG_KEY = 2;
+
+    static BytesReference fromConfigKey(String table, String column) {
+        try (BytesStreamOutput key = new BytesStreamOutput()) {
+            key.writeByte(FROM_CONFIG_KEY);
+            key.writeString(table);
+            key.writeString(column);
+            return key.bytes();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```

Comment:
The same, can we avoid catching this IOException?

Reply:
Checking.